drivers/mysql/lock.go

```diff
@@ -0,0 +1,194 @@
+//
+// A Mutex must not be copied after first use.
+type Mutex struct {
+	key string
+
+	// lock guards the variables used to manage the refresh task, and is not itself related to
+	// the db lock.
+	lock        sync.Mutex
+	stopRefresh chan bool
+	refreshDone chan bool
+	conn        *sql.Conn
+}
+
+// NewMutex creates a mutex with the given key name.
+//
+// returns error if key is empty.
+func NewMutex(key string, driver drivers.Driver) (*Mutex, error) {
+	key, err := drivers.MakeLockKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), drivers.TTL)
+	defer cancel()
+
+	ms, ok := driver.(*mysql)
+	if !ok {
+		return nil, errors.New("incorrect implementation of the driver")
+	}
+
+	createTableIfNotExistsQuery := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s (Id varchar(64) NOT NULL, ExpireAt bigint(20) NOT NULL, PRIMARY KEY (Id)) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4", drivers.MutexTableName)
+	if _, err = ms.conn.ExecContext(ctx, createTableIfNotExistsQuery); err != nil {
+		return nil, err
+	}
+
+	return &Mutex{
+		key:  key,
+		conn: ms.conn,
+	}, nil
+}
+
+// lock makes a single attempt to lock the mutex, returning true only if successful.
+func (m *Mutex) tryLock(ctx context.Context) (bool, error) {
+	now := time.Now()
+	query := fmt.Sprintf("INSERT INTO %s (Id, ExpireAt) VALUES ('%s', %d)", drivers.MutexTableName, m.key, now.Add(drivers.TTL).Unix())
+	if _, err := m.conn.ExecContext(ctx, query); err != nil {
+		err2 := m.releaseLock(ctx, now)
+		if err2 == nil {
+			// lock has been released due to expiration
+			return true, nil
+		}
+		return false, errors.Wrapf(err, "failed to lock mutex")
+	}
+
+	return true, nil
+}
+
+func (m *Mutex) releaseLock(ctx context.Context, t time.Time) error {
+	e, err := m.getExpireAt(ctx)
+	if err != nil {
+		return err
+	}
+
+	if t.Unix() < e {
+		return errors.New("could not release the lock")
+	}
+
+	query := fmt.Sprintf("UPDATE %s SET ExpireAt = %d WHERE Id = '%s'", drivers.MutexTableName, t.Add(drivers.TTL).Unix(), m.key)
+	_, err = m.conn.ExecContext(ctx, query)
+	if err != nil {
+		return errors.Wrap(err, "unable to set new expireat for mutex")
+	}
+	return nil
+}
+
+func (m *Mutex) getExpireAt(ctx context.Context) (int64, error) {
+	var expireAt int64
+	query := fmt.Sprintf("SELECT ExpireAt FROM %s WHERE Id = '%s'", drivers.MutexTableName, m.key)
+	err := m.conn.QueryRowContext(ctx, query).Scan(&expireAt)
+	if err != nil {
+		return -1, errors.Wrap(err, "failed to fetch mutex from db")
+	}
+
+	return expireAt, nil
+}
+
+// refreshLock rewrites the lock key value with a new expiry, returning nil only if successful.
+func (m *Mutex) refreshLock(ctx context.Context) error {
+	e, err := m.getExpireAt(ctx)
+	if err != nil {
+		return err
+	}
+
+	tmp := time.Unix(e, 0)
+	query := fmt.Sprintf("UPDATE %s SET ExpireAt = %d WHERE Id = '%s'", drivers.MutexTableName, tmp.Add(drivers.TTL).Unix(), m.key)
+	_, err = m.conn.ExecContext(ctx, query)
```

Comment:
We can do the update in a single operation itself. Or otherwise, this needs to be done inside a transaction. Otherwise, it runs into concurrency issues.